### PR TITLE
wip(events): Making `Saga` a domain model

### PR DESCRIPTION
**Pull request by jeyrschabu**
- adding internal events

### CODE REVIEW

File: clouddriver-saga/src/main/kotlin/com/netflix/spinnaker/clouddriver/saga/models/Saga.kt [hunk 1]
```diff
@@ -68,6 +78,20 @@ class Saga(
   var compensating: Boolean = false
```
Comment (jeyrschabu) [inline clouddriver-saga/src/main/kotlin/com/netflix/spinnaker/clouddriver/saga/models/Saga.kt]: Wondering if we should add a default compensate in SagaEventHandler and have that one just turn the flag to true. Then have impl have to call `super.compensate`. Has to be a better way

File: clouddriver-saga/src/main/kotlin/com/netflix/spinnaker/clouddriver/saga/events.kt [hunk 2]
```diff
@@ -60,6 +60,42 @@ class SagaLogAppended(
...
+  sagaId: String,
+  val oldValue: Long,
+  val newValue: Long
+) : SagaEvent(sagaName, sagaId)
+
+class SagaRequiredEventsAdded(
+  sagaName: String,
+  sagaId: String,
+  val eventNames: List<String>,
+  val currentNumOfEvents: Int
```
Comment (robzienert) [inline clouddriver-saga/src/main/kotlin/com/netflix/spinnaker/clouddriver/saga/events.kt]: I don't see any value in tracking `currentNumEvents`.

File: clouddriver-saga/src/main/kotlin/com/netflix/spinnaker/clouddriver/saga/events.kt [hunk 3]
```diff
@@ -60,6 +60,42 @@ class SagaLogAppended(
...
+) : SagaEvent(sagaName, sagaId)
+
+class SagaCompleted(
+  sagaName: String,
+  sagaId: String
+) : SagaEvent(sagaName, sagaId)
```
Comment (robzienert) [inline clouddriver-saga/src/main/kotlin/com/netflix/spinnaker/clouddriver/saga/events.kt]: We'll want a `val success: Boolean`. `SagaCompleted` will be fired at the completion of the compensation lifecycle as well, in which case the success value would be `false`.

File: clouddriver-saga/src/main/kotlin/com/netflix/spinnaker/clouddriver/saga/models/Saga.kt [hunk 4]
```diff
@@ -60,6 +66,10 @@ class Saga(
   var completed: Boolean = false
     set(value) {
       if (value != field) {
+        addEvent(SagaCompleted(
+          name,
+          id
+        ))
         dirty = true
```
Comment (robzienert) [inline clouddriver-saga/src/main/kotlin/com/netflix/spinnaker/clouddriver/saga/models/Saga.kt]: Get rid of `dirty` everywhere.

File: clouddriver-saga/src/main/kotlin/com/netflix/spinnaker/clouddriver/saga/models/Saga.kt (hunk 1 above)
Comment (robzienert) [reply clouddriver-saga/src/main/kotlin/com/netflix/spinnaker/clouddriver/saga/models/Saga.kt]: What do you mean by "default compensate"? How could there be a generic default?

**Comment by jeyrschabu** [comment]
Closing this guy. Destroyed my local branch, going to start fresh in another PR coming soon

**Pull request closed without merging**